### PR TITLE
lfr cleanup and update.

### PR DESCRIPTION
**Pull request by pronics2004**
Removed numba dependency and cleaned up the code for consistent terminology.

This implementation is closer to vanilla LFR, where the feature distance computation is standard l2-distance without any learned scaling parameters (`alpha`).

### CODE REVIEW

File: aif360/algorithms/preprocessing/lfr.py [hunk 1]
```diff
@@ -67,6 +67,8 @@ def __init__(self,
         self.print_interval = print_interval
         self.verbose = verbose
 
+        self.w = None
+        self.prototypes = None
         self.learned_model = None
 
     def fit(self, dataset, maxiter=5000, maxfun=5000, **kwargs):
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/algorithms/preprocessing/lfr.py:74]: also, I just noticed we need to add descriptions for maxiter and maxfun to the docstring

File: aif360/algorithms/preprocessing/lfr.py [hunk 2]
```diff
@@ -81,30 +83,35 @@ def fit(self, dataset, maxiter=5000, maxfun=5000, **kwargs):
...
 
-        self.learned_model = optim.fmin_l_bfgs_b(lfr_helpers.LFR_optim_obj, x0=model_inits, epsilon=1e-5,
-                                  args=(training_sensitive, training_nonsensitive,
-                                        ytrain_sensitive[:, 0], ytrain_nonsensitive[:, 0], self.k, self.Ax,
-                                        self.Ay, self.Az, 0, self.print_interval),
-                                  bounds=bnd, approx_grad=True, maxfun=maxfun,
-                                  maxiter=maxiter, disp=self.verbose)[0]
         return self
 
     def transform(self, dataset, threshold=0.5, **kwargs):
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr.py:117]: same thing here: can we remove `**kwargs`?

File: aif360/algorithms/preprocessing/lfr.py [hunk 3]
```diff
@@ -81,30 +83,35 @@ def fit(self, dataset, maxiter=5000, maxfun=5000, **kwargs):
...
-        ytrain_sensitive = dataset.labels[sensitive_idx]
-        ytrain_nonsensitive = dataset.labels[nonsensitive_idx]
-
-        model_inits = np.random.uniform(size=self.features_dim * 2 + self.k + self.features_dim * self.k)
+        unprivileged_sample_ids = np.array(np.where(protected_attributes == self.unprivileged_group_protected_attribute_value))[0].flatten()
+        privileged_sample_ids = np.array(np.where(protected_attributes == self.privileged_group_protected_attribute_value))[0].flatten()
+        features_unpriveleged = dataset.features[unprivileged_sample_ids]
+        features_priveleged = dataset.features[privileged_sample_ids]
+        labels_unpriveleged = dataset.labels[unprivileged_sample_ids]
+        labels_priveleged = dataset.labels[privileged_sample_ids]
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr.py:94]: just a typo: can we rename these to `*privileged`

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 4]
```diff
@@ -1,109 +1,46 @@
...
-        yhat[i] = 1e-6 if yhat[i] <= 0 else yhat[i]
-        yhat[i] = 0.999 if yhat[i] >= 1 else yhat[i]
-        L_y += -1 * y[i] * np.log(yhat[i]) - (1.0 - y[i]) * np.log(1.0 - yhat[i])
-    return yhat, L_y
 
-@jit
-def LFR_optim_obj(params, data_sensitive, data_nonsensitive, y_sensitive,
-                  y_nonsensitive, k=10, A_x = 0.01, A_y = 0.1, A_z = 0.5, results=0, print_inteval=250):
+def LFR_optim_objective(parameters, x_unpriveleged, x_priveleged, y_unpriveleged,
+                        y_priveleged, k=10, A_x=0.01, A_y=0.1, A_z=0.5, print_inteval=250):
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr_helpers/helpers.py:9]: same here: `priveleged` -> `privileged`
also, `inteval` -> `interval`

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 5]
```diff
@@ -1,109 +1,46 @@
...
-    x_n_hat_nonsensitive, L_x2 = x_n_hat(data_nonsensitive, M_nk_nonsensitive, v, Nns, P, k)
-    L_x = L_x1 + L_x2
+    total_loss = A_x * L_x + A_y * L_y + A_z * L_z
 
-    yhat_sensitive, L_y1 = yhat(M_nk_sensitive, y_sensitive, w, Ns, k)
-    yhat_nonsensitive, L_y2 = yhat(M_nk_nonsensitive, y_nonsensitive, w, Nns, k)
-    L_y = L_y1 + L_y2
+    if LFR_optim_objective.steps % print_inteval == 0:
+        print("step: {}, loss: {}, L_x: {},  L_y: {},  L_z: {}".format(
+            LFR_optim_objective.steps, total_loss, L_x,  L_y,  L_z))
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr_helpers/helpers.py:32]: if `verbose=0` in LFR, we shouldn't print this

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 6]
```diff
@@ -1,109 +1,46 @@
...
-    else:
-        return criterion
-LFR_optim_obj.iters = 0
+def get_xhat_y_hat(prototypes, w, x):
+    M = softmax(cdist(x, prototypes), axis=1)
+    x_hat = np.matmul(M, prototypes)
+    y_hat = np.clip(
+        np.matmul(M, w.reshape((-1, 1))),
+        1e-6,
+        0.999
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr_helpers/helpers.py:44]: these values seem sort of arbitrary -- was there a reason for choosing these as opposed to machine epsilon or something?

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 7]
```diff
@@ -1,109 +1,46 @@
...
-    if LFR_optim_obj.iters % print_inteval == 0:
-        print(LFR_optim_obj.iters, criterion)
 
-    if results:
-        return yhat_sensitive, yhat_nonsensitive, M_nk_sensitive, M_nk_nonsensitive
-    else:
-        return criterion
-LFR_optim_obj.iters = 0
+def get_xhat_y_hat(prototypes, w, x):
+    M = softmax(cdist(x, prototypes), axis=1)
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr_helpers/helpers.py:39]: shouldn't this be `softmax(-cdist(...))`?

File: aif360/algorithms/preprocessing/lfr.py [hunk 8]
```diff
@@ -179,4 +160,42 @@ def transform(self, dataset, threshold=0.5, **kwargs):
...
+
+        metric_transf_train = BinaryLabelDatasetMetric(dataset_transf_train,
+                                                       unprivileged_groups=unprivileged_groups,
+                                                       privileged_groups=privileged_groups)
+        print("Classification threshold = %f" % threshold)
+        print(classification_report(dataset_orig_train.labels, dataset_transf_train.labels))
+        print(
+            "Difference in mean outcomes between unprivileged and privileged groups = %f" % metric_transf_train.mean_difference())
+        print("Consistency of labels in transformed training dataset= %f" % metric_transf_train.consistency())
+        print("Consistency of labels in original training dataset= %f" % metric_orig_train.consistency())
```
[the remark this answers is not in the document]
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: done.

File: aif360/algorithms/preprocessing/lfr.py [hunk 9]
```diff
@@ -179,4 +160,42 @@ def transform(self, dataset, threshold=0.5, **kwargs):
     def fit_transform(self, dataset, seed=None):
         """fit and transform methods sequentially"""
 
-        return self.fit(dataset, seed=seed).transform(dataset)
\ No newline at end of file
+        return self.fit(dataset, seed=seed).transform(dataset)
```
[the remark this answers is not in the document]
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: fixed.

File: aif360/algorithms/preprocessing/lfr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: thanks. fixed now.

File: aif360/algorithms/preprocessing/lfr.py (hunk 3 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: thanks for catching this. fixed.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py (hunk 5 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py]: yes. fixed.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py (hunk 7 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py]: good catch. fixed now.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 10]
```diff
@@ -38,6 +38,7 @@ def LFR_optim_objective(parameters, x_unprivileged, x_privileged, y_unprivileged
 def get_xhat_y_hat(prototypes, w, x):
     M = softmax(-cdist(x, prototypes), axis=1)
     x_hat = np.matmul(M, prototypes)
+    w = np.clip(w, np.finfo(float).eps, 1.0 - np.finfo(float).eps)
```
Comment (pronics2004) [inline aif360/algorithms/preprocessing/lfr_helpers/helpers.py:41]: Constraining the wrights between 0 and 1.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py (hunk 6 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py]: fixed.

File: aif360/algorithms/preprocessing/lfr.py (hunk 2 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: done

File: aif360/algorithms/preprocessing/lfr.py [hunk 11]
```diff
@@ -81,33 +85,38 @@ def fit(self, dataset, maxiter=5000, maxfun=5000, **kwargs):
...
         bnd = []
-        for i, _ in enumerate(model_inits):
+        for i, _ in enumerate(parameters_initialization):
             if i < self.features_dim * 2 or i >= self.features_dim * 2 + self.k:
                 bnd.append((None, None))
             else:
                 bnd.append((0, 1))
+        lfr_helpers.LFR_optim_objective.steps = 0
+
+        self.learned_model = optim.fmin_l_bfgs_b(lfr_helpers.LFR_optim_objective, x0=parameters_initialization, epsilon=1e-5,
```
Comment (nrkarthikeyan) [inline aif360/algorithms/preprocessing/lfr.py:108]: indentation issue here

File: examples/demo_lfr.ipynb [hunk 12]
```diff
@@ -30,9 +30,13 @@
     "from sklearn.linear_model import LogisticRegression\n",
     "from sklearn.preprocessing import StandardScaler\n",
     "from sklearn.metrics import accuracy_score\n",
+    "from sklearn.metrics import classification_report\n",
```
Comment (nrkarthikeyan) [inline examples/demo_lfr.ipynb:33]: Please bring the reference to the top of the notebook

File: aif360/algorithms/preprocessing/lfr.py [hunk 13]
```diff
@@ -179,4 +160,42 @@ def transform(self, dataset, threshold=0.5, **kwargs):
     def fit_transform(self, dataset, seed=None):
         """fit and transform methods sequentially"""
 
-        return self.fit(dataset, seed=seed).transform(dataset)
\ No newline at end of file
+        return self.fit(dataset, seed=seed).transform(dataset)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: this is fixed now.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py [hunk 14]
```diff
@@ -1,109 +1,46 @@
 # Based on code from https://github.com/zjelveh/learning-fair-representations
```
[the remark this answers is not in the document]
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py:1]: yes still based on it.

File: examples/demo_lfr.ipynb (hunk 12 above)
Comment (pronics2004) [reply examples/demo_lfr.ipynb:33]: ok

File: aif360/algorithms/preprocessing/lfr.py [hunk 15]
```diff
@@ -81,33 +85,38 @@ def fit(self, dataset, maxiter=5000, maxfun=5000, **kwargs):
...
+
+        # Initialize the LFR optim objective parameters
+        parameters_initialization = np.random.uniform(size=self.k + self.features_dim * self.k)
         bnd = []
-        for i, _ in enumerate(model_inits):
+        for i, _ in enumerate(parameters_initialization):
             if i < self.features_dim * 2 or i >= self.features_dim * 2 + self.k:
                 bnd.append((None, None))
             else:
                 bnd.append((0, 1))
```
Comment (hoffmansc) [inline aif360/algorithms/preprocessing/lfr.py:105]: I think this should be: `bnd = [(0, 1)]*self.k + [(None, None)]*self.features_dim*self.k` so w is bound to (0, 1) and the prototypes are unbounded.

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py (hunk 10 above)
Comment (hoffmansc) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py:41]: with the bnd fix above, I think this becomes unnecessary

File: aif360/algorithms/preprocessing/lfr.py (hunk 15 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr.py]: yes. fixed,

File: aif360/algorithms/preprocessing/lfr_helpers/helpers.py (hunk 10 above)
Comment (pronics2004) [reply aif360/algorithms/preprocessing/lfr_helpers/helpers.py]: fixed.

**Comment by hoffmansc** [comment]
> We can also get rid of the `except` part of the import in lfr.py (lines 7-10).

@pronics2004 other than this, looks great. Thanks!

**Comment by pronics2004** [comment]
> > We can also get rid of the `except` part of the import in lfr.py (lines 7-10).
> 
> @pronics2004 other than this, looks great. Thanks!

Done. Thanks!
